File: Examples/Examples/ListView/ListViewPullToRefresh.cs
using System;
using System.Collections;
using System.Collections.Generic;

using Foundation;
using UIKit;
using CoreFoundation;

using TelerikUI;

namespace Examples
{
	[Register("ListViewPullToRefresh")]
	public class ListViewPullToRefresh: XamarinExampleViewController
	{
		List<string> data = new List<string> ();
		TKDataSource dataSource = new TKDataSource();
		ListViewDataSource listViewDataSource;
		ListViewDelegate listViewDelegate;
		int newItemsCount = 0;

		public override void ViewDidLoad ()
		{
			base.ViewDidLoad ();

			dataSource.LoadDataFromJSONResource ("ListViewSampleData", "json", "teams");
			dataSource.GroupItemSourceKey = "items";
			dataSource.FilterWithQuery ("key like 'Marketing'");
			dataSource.GroupWithKey ("key");

			this.UpdateData (3);

			this.listViewDataSource =  new ListViewDataSource (this);
			this.listViewDelegate = new ListViewDelegate (this);

			TKListView listView = new TKListView (this.View.Bounds);
			listView.RegisterClassForCell (new ObjCRuntime.Class (typeof(TKListViewCell)), "cell");
			listView.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
			listView.DataSource = this.listViewDataSource;
			listView.Delegate = this.listViewDelegate;

			// >> listview-pull-to-refresh-cs
			listView.AllowsPullToRefresh = true;
			// << listview-pull-to-refresh-cs

			listView.PullToRefreshTreshold = 70;
			listView.PullToRefreshView.BackgroundColor = UIColor.Blue;
			listView.PullToRefreshView.ActivityIndicator.Color = UIColor.White;
			this.View.AddSubview (listView);
		}

		int UpdateData(int count)
		{
			TKDataSourceGroup group = (TKDataSourceGroup)this.dataSource.Items [0];
			long startIndex = this.data.Count;
			int i = 0;
			for (; i < count; i++) {
				if (i + startIndex >= group.Items.Length) {
					return i;
				}

				Random r = new Random ();
				int points = r.Next (0, 101);
				this.data.Insert (0, string.Format ("{0}: {1} points", group.Items [i + startIndex], points));
			}

			return i;
		}

		bool IsUpdated(NSIndexPath indexPath)
		{
			return indexPath.Row < newItemsCount;
		}

		public class ListViewDataSource: TKListViewDataSource
		{
			ListViewPullToRefresh owner;

			public ListViewDataSource(ListViewPullToRefresh owner)
			{
				this.owner = owner;
			}

			public override int NumberOfItemsInSection (TKListView listView, int section)
			{
				return this.owner.data.Count;
			}

			public override TKListViewCell CellForItem (TKListView listView, NSIndexPath indexPath)
			{
				TKListViewCell cell = listView.DequeueReusableCell ("cell", indexPath) as TKListViewCell;
				bool isUpdated = this.owner.IsUpdated (indexPath);
				cell.TextLabel.Text = this.owner.data [indexPath.Row];
				cell.BackgroundView.BackgroundColor = isUpdated ? new UIColor (1f, 1f, 0f, 0.4f) : UIColor.White;
				if (isUpdated) {
					UIView.Animate (0.5f, () => {
						cell.BackgroundView.Alpha = 1;
					});
				}

				return cell;
			}
		}

		class ListViewDelegate: TKListViewDelegate 
		{
			ListViewPullToRefresh owner;

			public ListViewDelegate(ListViewPullToRefresh owner) 
			{
				this.owner = owner;
			}

			public override void DidPull (TKListView listView, nfloat offset)
			{
				listView.PullToRefreshView.Alpha = (float)Math.Min (offset / listView.PullToRefreshTreshold, 1.0f);
			}

			// >> listview-should-refresh-cs
			public override bool ListViewShouldRefreshOnPull (TKListView listView)
			{
				DispatchQueue.DefaultGlobalQueue.DispatchAsync (() => {
					Random r = new Random();
					this.owner.newItemsCount = this.owner.UpdateData(1 + r.Next(0, 4));
					DispatchQueue.MainQueue.DispatchAfter(new DispatchTime(DispatchTime.Now, 2 * 500000000), () => {
						listView.DidRefreshOnPull();
						if (this.owner.newItemsCount < 1) {
							UIAlertView alert = new UIAlertView("Pull to refresh", "No more data available!",null,"Close",null);
							alert.Show();
						}
					});
				});

				return true;
			}
			// << listview-should-refresh-cs
		}
	}
}

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/ListViewPullToRefresh.swift
//
//  ListViewPullToRefresh.swift
//  TelerikUIExamplesInSwift
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

import UIKit

class ListViewPullToRefresh: TKExamplesExampleViewController, TKListViewDataSource, TKListViewDelegate {

    let dataSource = TKDataSource()
    var data = [String]()
    var newItemsCount : Int = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.navigationController?.navigationBar.translucent = false
        
        // Do any additional setup after loading the view.
        dataSource.loadDataFromJSONResource("ListViewSampleData", ofType:"json", rootItemKeyPath: "teams")
        dataSource.groupItemSourceKey = "items"
        dataSource.filterWithQuery("key like 'Marketing'")
        dataSource.groupWithKey("key")
        self.updateData(3)
        
        let listView = TKListView(frame: self.view.bounds)
        listView.autoresizingMask = UIViewAutoresizing(rawValue: UIViewAutoresizing.FlexibleWidth.rawValue | UIViewAutoresizing.FlexibleHeight.rawValue)
        listView.dataSource = self
        listView.delegate = self
        
        // >> listview-pull-to-refresh-swift
        listView.allowsPullToRefresh = true
        // << listview-pull-to-refresh-swift
        
        listView.pullToRefreshTreshold = 70
        listView.pullToRefreshView.backgroundColor = UIColor.blueColor()
        listView.pullToRefreshView.activityIndicator.color = UIColor.whiteColor()
        listView.registerClass(TKListViewCell.classForCoder() , forCellWithReuseIdentifier: "cell")
        self.view.addSubview(listView)
    }

    override func willMoveToParentViewController(parent: UIViewController?) {
        if parent == nil {
            self.navigationController?.navigationBar.translucent = true
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func updateData(count: NSInteger) -> Int {
        let group = dataSource.items[0] as! TKDataSourceGroup
        let startIndex = data.count
        var j = 0
        for i in 0..<count {
            if(i + startIndex >= group.items.count){
                return i
            }
            let points = arc4random() % 100
            data.insert("\(group.items[i + startIndex]) \(points) points", atIndex: 0)
            j+=1
         }
         return j
    }
    
    func isUpdated(indexPath: NSIndexPath) -> Bool {
      return indexPath.row < newItemsCount
    }
    
// MARK: - TKListViewDataSource
    
    func listView(listView: TKListView, numberOfItemsInSection section: Int) -> Int {
        return data.count
    }
    
    func listView(listView: TKListView, cellForItemAtIndexPath indexPath: NSIndexPath) -> TKListViewCell? {
        let cell = listView.dequeueReusableCellWithReuseIdentifier("cell", forIndexPath: indexPath) as! TKListViewCell
        let isUpdated = self.isUpdated(indexPath)
        cell.textLabel.text = data[indexPath.row]
        let backgroundView : UIView = cell.backgroundView!
        backgroundView.backgroundColor = isUpdated ? UIColor(red: 1.0, green: 1.0, blue: 0.0, alpha: 0.4): UIColor.whiteColor()
        backgroundView.alpha = isUpdated ? 0.0 : 1.0
        if(isUpdated){
           UIView.animateWithDuration(NSTimeInterval(0.5), animations: { () -> Void in
               backgroundView.alpha = 1
           })
        }
        return cell
    }
    
// MARK: - TKListViewDelegate
    
    func listView(listView: TKListView, didPullWithOffset offset: CGFloat) {
        listView.pullToRefreshView.alpha = min(offset/listView.pullToRefreshTreshold, 1.0)
    }
    
    // >> listview-should-refresh-swift
    func listViewShouldRefreshOnPull(listView: TKListView) -> Bool {
        
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), {
            
            self.newItemsCount = self.updateData(1 + Int(arc4random() % 3))
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(2 * Double(NSEC_PER_SEC))), dispatch_get_main_queue(), { () -> Void in
                listView.didRefreshOnPull()
                
                if(self.newItemsCount < 1){
                    let alert = TKAlert()
                    alert.title = "Pull to refresh"
                    alert.message = "No more data available"
                    alert.addAction(TKAlertAction(title: "Close", handler: { (alert:TKAlert, action:TKAlertAction) -> Bool in
                        return true
                    }))
                    alert.show(true)
                }
            })
        })
        return true
    }
    // << listview-should-refresh-swift
}
